File: NutriSmart/app/src/main/java/com/example/nutrismart/ui/screens/home/HomeUiState.kt
package com.example.nutrismart.ui.screens.home

import com.example.nutrismart.data.remote.MealDto
import com.example.nutrismart.data.remote.ShoppingListDto
import java.time.LocalTime

data class HomeUiState(
    val isLoading: Boolean = true,
    val isSwapping: Boolean = false,
    val steps: Int = 0,
    val stepsGoal: Int = 10000,
    val waterConsumedMl: Int = 0,
    val glassSizeMl: Int = 250,
    val waterGoalMl: Int = 2000,
    val breakfast: MealDto? = null,
    val lunch: MealDto? = null,
    val dinner: MealDto? = null,
    val snack: MealDto? = null,
    val bonusSnack: MealDto? = null,
    val availableBonusSnacks: List<MealDto> = emptyList(),
    val shoppingList: ShoppingListDto? = null,
    val alternatives: List<MealDto> = emptyList(),
    val wakeUpTime: LocalTime = LocalTime.of(8,0),
    val burnedCalories: Int = 0,
    val weight: Double = 0.0
) {
    val totalCaloriesGoal: Int get() = listOfNotNull(breakfast, lunch, dinner, snack).sumOf { it.calories }.coerceAtLeast(1)
    val carbsGoal: Int get() = listOfNotNull(breakfast, lunch, dinner, snack).sumOf { it.carbs }.coerceAtLeast(1)
    val proteinGoal: Int get() = listOfNotNull(breakfast, lunch, dinner, snack).sumOf { it.protein }.coerceAtLeast(1)
    val fatGoal: Int get() = listOfNotNull(breakfast, lunch, dinner, snack).sumOf { it.fat }.coerceAtLeast(1)

    val caloriesConsumed: Int get() = listOfNotNull(breakfast, lunch, dinner, snack, bonusSnack).filter { it.consumed }.sumOf { it.calories }
    val carbsConsumed: Int get() = listOfNotNull(breakfast, lunch, dinner, snack, bonusSnack).filter { it.consumed }.sumOf { it.carbs }
    val proteinConsumed: Int get() = listOfNotNull(breakfast, lunch, dinner, snack, bonusSnack).filter { it.consumed }.sumOf { it.protein }
    val fatConsumed: Int get() = listOfNotNull(breakfast, lunch, dinner, snack, bonusSnack).filter { it.consumed }.sumOf { it.fat }
    val caloriesRemaining: Int get() = totalCaloriesGoal - caloriesConsumed + burnedCalories
}

File: NutriSmartIOS/ui/screens/home/HomeUiState.swift
//
//  HomeUiState.swift
//  NutriSmartIOS
//

import Foundation

struct HomeUiState {
    var isLoading: Bool = false
    var weight: Double = 0.0
    var steps: Int = 0
    var burnedCalories: Int = 0
    var isImperial: Bool = false
    
    var breakfast: MealDto? = nil
    var lunch: MealDto? = nil
    var dinner: MealDto? = nil
    var snack: MealDto? = nil
    
    var waterConsumedMl: Int = 0
    var waterGoalMl: Int = 2000
    var glassSizeMl: Int = 250
    
    var shoppingList: ShoppingListDto? = nil
    var alternatives: [MealDto] = []
    var isSwapping: Bool = false
    
    var wakeUpTime: Date = Calendar.current.date(bySettingHour: 8, minute: 0, second: 0, of: Date()) ?? Date()
    
    var bonusSnack: MealDto? = nil
    var availableBonusSnacks: [MealDto] = []
}
